src/webhook_launcher/app/models.py

```diff
@@ -392,24 +399,119 @@ def clean(self, exclude=None):
                     (project, self.user)
                 )
 
-    def trigger_build(self):
-        if not self.lsr:
-            mylsr, created = LastSeenRevision.objects.get_or_create(
-                mapping=self,
+    def trigger_build(self, user=None, tag=None, force=False):
+
+        # Only fire for projects which allow webhooks. We can't just
+        # rely on validation since a Project may forbid hooks after
+        # the hook was created
+        if self.project_disabled:
+            print "Project has build disabled"
+            return
+
+        handled = self.lsr.handled and self.lsr.tag == tag and not force
```

Comment:
Added check that the WebHookMapping is saved before calling trigger_build or handle_commit. This ensures that it gets a LastSeenRevision object via lsr property. Whether the lsr has proper values can affect the build result, but checking that is out of scope for this task. For that need to think about the different use cases and what we want to allow or not.